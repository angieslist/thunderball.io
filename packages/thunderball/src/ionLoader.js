import fs from 'fs';
import path from 'path';

module.exports = {
  // Get all folders in the 'ions' folder that contain a manifest file
  // and map them to an object with the path and loaded configuration
  load: (ionsDir, manifestName = 'manifest.js') => {
    if (!ionsDir) {
      throw new Error('No path for ions is specified');
    }

    try {
      return fs.readdirSync(ionsDir)
        .filter(file => fs.statSync(path.join(ionsDir, file)).isDirectory())
        .filter(folder => fs.existsSync(path.join(ionsDir, folder, manifestName)))
        .map((folder) => {
          const ion = {
            dir: folder,
            config: require(path.join(ionsDir, folder, manifestName)) || {},
          };
          // If ion has no 'name', then assume the folder name
          if (!ion.name) {
            ion.name = folder;
          }
          return ion;
        })
      ;
    } catch (e) {
      console.log(`Could not load ion directory: "${ionsDir}", using no ions...`);
      return [];
    }
  },
};
