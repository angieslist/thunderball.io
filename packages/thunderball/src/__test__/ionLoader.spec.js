import path from 'path';
import ionLoader from '../ionLoader';

describe('ionLoader.load', () => {
  it('loads all ions in an ions directory', () => {
    const ionsDir = path.resolve(__dirname, '../../../thunderball-hello-world/src/ions');
    const ions = ionLoader.load(ionsDir);
    expect(ions).toMatchSnapshot();
  });
});
