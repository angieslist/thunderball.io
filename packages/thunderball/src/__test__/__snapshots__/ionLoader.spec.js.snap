// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ionLoader.load loads all ions in an ions directory 1`] = `
Array [
  Object {
    "config": Object {
      "browser": Object {
        "page": Object {
          "createRoutes": "./createRoutes",
          "injectors": Array [
            "../../appInjectors",
          ],
          "path": "/*",
        },
      },
      "name": "home",
    },
    "dir": "home",
    "name": "home",
  },
  Object {
    "config": Object {
      "browser": Object {
        "page": Object {
          "createRoutes": "./createRoutes",
          "injectors": Array [
            "../../appInjectors",
            "./injectors",
          ],
          "path": "/todos*",
          "props": Object {
            "basePath": "/todos",
          },
        },
      },
      "name": "todos",
    },
    "dir": "todos",
    "name": "todos",
  },
  Object {
    "config": Object {
      "browser": Object {
        "page": Object {
          "createRoutes": "./createRoutes",
          "injectors": Array [
            "../../appInjectors",
            "./injectors",
          ],
          "path": "/todo-answers*",
          "props": Object {
            "basePath": "/todo-answers",
          },
        },
      },
      "middleware": "./api/middleware",
      "name": "todos-answers",
    },
    "dir": "todos-answers",
    "name": "todos-answers",
  },
]
`;
